Allow to use github.com/Code-Hex/vz/v2 from macOS 11/macOS 10

cfergeau opened the pull request:
This PR adds version annotation to the objective C API so that it's possible to
build vz/use it on a macOS 10/11 machine even if it does not implement all the
latest virtualization.framework API. When trying to use unsupported API,
vz.ErrUnsupportedOSVersion will be returned.

This addresses https://github.com/Code-Hex/vz/issues/41 as this
allows to support macOS 10, macOS 11 and macOS 12 in the same build.

cfergeau commented:
I forgot to mention that after these changes there is still these warnings when building:
```
$ make
go build -o virtualization .
# github.com/Code-Hex/vz/v2
virtualization_view.m:170:5: warning: 'VZVirtualMachineView' is only available on macOS 12.0 or newer [-Wunguarded-availability-new]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtualMachineView.h:23:12: note: 'VZVirtualMachineView' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
../../virtualization_view.h:25:12: note: annotate 'AppDelegate' with an availability attribute to silence this warning
virtualization_view.m:184:5: warning: 'VZVirtualMachineView' is only available on macOS 12.0 or newer [-Wunguarded-availability-new]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtualMachineView.h:23:12: note: 'VZVirtualMachineView' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
virtualization_view.m:184:5: note: enclose 'VZVirtualMachineView' in an @available check to silence this warning
virtualization_view.m:184:37: warning: 'VZVirtualMachineView' is only available on macOS 12.0 or newer [-Wunguarded-availability-new]
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtualMachineView.h:23:12: note: 'VZVirtualMachineView' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
virtualization_view.m:184:37: note: enclose 'VZVirtualMachineView' in an @available check to silence this warning
codesign --entitlements vz.entitlements -s - ./virtualization
./virtualization: is already signed
make: *** [codesign] Error 1
```

I'd need help to silence them as I don't really manage to read the related objC code :(

Code-Hex commented on audio.go:
```diff
@@ -1,7 +1,7 @@
 
 /*
-#cgo darwin CFLAGS: -x objective-c -fno-objc-arc
+#cgo darwin CFLAGS: -x objective-c -fno-objc-arc -mmacosx-version-min=10.13
```
Please remove this flag from each file
```suggestion
#cgo darwin CFLAGS: -x objective-c -fno-objc-arc
```

cfergeau replied:
Is it fine if it's `-mmacosx-version-min=10.13`? Or do you prefer if I set it to `CGO_LDFLAGS` in the Makefile?

cfergeau replied:
`-mmacosx-version-min=11.0` is needed in order to generate binaries which will build on older macOS versions. If it's not added to the go files, I think users of Code-Hex/vz will have to add it themselves to their build flags. Do you prefer to leave this choice to users of the module?

Code-Hex replied:
Yes. Because I referred about the version should use above than 11.x in readme. So I don't think this user is used in older version

Code-Hex commented:
@cfergeau Thanks!
Could you rebase some commit logs? and please fix it also for 11.0 APIs similar to your fixes. Thanks

Code-Hex commented:
Please discard this commit
https://github.com/Code-Hex/vz/pull/48/commits/2b0ec585be1d073ee6b4ae644818d66b32922afc

cfergeau commented:
> Please discard this commit [2b0ec58](https://github.com/Code-Hex/vz/commit/2b0ec585be1d073ee6b4ae644818d66b32922afc)

Apologies for this, I mistakenly pushed some commits by mistake, the PR should have stopped at `stop/canStop: Return errors from go API`. It's now removed.

cfergeau commented:
> @cfergeau Thanks! and please fix it also for 11.0 APIs similar to your fixes. Thanks

> please fix it also for 11.0 APIs similar to your fixes. 

This was the plan after checking with this PR that the approach I took is acceptable, I'll work on it now!

> Could you rebase some commit logs? 

What commit logs do you want rebased? I removed some of the commits I added by mistake, do you want more changes on logs of the existing commits?

Code-Hex commented:
@cfergeau It's ok. Please continue.

Please write some test code. Thank you!

cfergeau commented:
I've added some test to this series. It's still macos11 only, macos10 is progressing well in https://github.com/cfergeau/vz/tree/macos10, it's missing:
- commit logs
- tests

Tests are slightly more complicated, as the test must be built on macOS11 (no Virtualization.h header on macOS10 machines), and then run on macOS10. I haven't looked yet into doing that in github actions. I can't manually test as I no longer have a macOS10 machine.

cfergeau commented:
I've added macos 10 "support". I've also added some tests. The macos10 tests are not run automatically in gh actions

Code-Hex commented:
@cfergeau Could you please remove the part about the test code once you have written it, because to be honest the test code you write is unfortunately not of mergeable quality?

cfergeau commented:
> @cfergeau Could you please remove the part about the test code once you have written it, because to be honest the test code you write is unfortunately not of mergeable quality?

You asked about adding tests ( https://github.com/Code-Hex/vz/pull/48#issuecomment-1251144329 ) so I wrote them. They test if we get ErrUnsupportedOS when expected.
If you prefer, yes I can remove them. Is it only the tests which are not of mergeable quality? Or are there more issues I need to fix?

Code-Hex commented:
@cfergeau A test to determine availability can be done by mocking the version information.

Revert the test code, actions settings, and makefile modifications

Code-Hex commented on virtualization.go:
```diff
@@ -294,3 +348,23 @@ func (v *VirtualMachine) Stop(fn func(error)) {
+		return 0
+	}
+	osverArray := strings.Split(osver, ".")
+	if len(osverArray) < 1 {
+		return 0
+	}
```
Please remove this handling because the result will be more than or equal 1
https://go.dev/play/p/BJBkxhG0-lQ

cfergeau commented on virtualization.go:
```diff
@@ -294,3 +348,23 @@ func (v *VirtualMachine) Stop(fn func(error)) {
 func (v *VirtualMachine) StartGraphicApplication(width, height float64) {
 	C.startVirtualMachineWindow(v.Ptr(), C.double(width), C.double(height))
 }
+
```
Thanks for the suggestion, I will do this!

cfergeau replied:
Latest push should be closer to what you want. I kept the build tags. If you prefer, I can set `majorVersion` at the beginning of each test.

Code-Hex commented:
Why can't you do what I commented on?

> Revert the test code, actions settings, and makefile modifications

cfergeau commented:
> Why can't you do what I commented on?
> 
> > Revert the test code, actions settings, and makefile modifications

I've removed all of this in my latest push. I thought the addition of `fetchMajorVersion` was a way of making the tests acceptable.
To be honest, I much prefer to have sub-optimal tests. It is always good to have some form of test code. Then the tests can be fixed to your liking/removed later.

Code-Hex reviewed (changes requested):
Probably last requests 🙏

cfergeau commented on virtualization.go:
```diff
@@ -198,6 +211,9 @@ func (v *VirtualMachine) CanRequestStop() bool {
+//
+// This is only supported on macOS 12 and newer, false will always be returned
+// on older versions.
 func (v *VirtualMachine) CanStop() bool {
```
I will add some changes to `CanStop`/... to handle older macOS from the go code.

Regarding your exception suggestion, the `@available` check has to be `if (@available(...)`, it cannot be negated, see https://stackoverflow.com/questions/46388490/negating-objective-cs-available-keyword , so this would become 
```
if (@available(macOS 12, *)) {
} else {
  [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise];
}
// write original code here.
```

I think I'll wrap the code above in a macro. It does not look very nice ^^

```
#define THROW_IF_MACOS_OLDER_THAN(version) \
    if (@available(macOS version, *)) { \
    } else { \
        [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise]; \
    }
```

and then 
```
THROW_IF_MACOS_OLDER_THAN(12)
// write original code here.
```

Code-Hex replied:
LGTM :+1:

cfergeau replied:
There is one big limitation with this. If I use this code:
```
void setDirectorySharingDevicesVZVirtualMachineConfiguration(void *config, void *directorySharingDevices)
{
    THROW_IF_MACOS_OLDER_THAN(12)

    [(VZVirtualMachineConfiguration *)config setDirectorySharingDevices:[(NSMutableArray *)directorySharingDevices copy]];
}
```

Then when building for macOS 11 (`CGO_CFLAGS=-mmacosx-version-min=11.0`) I get this warning:
```
virtualization.m:277:46: warning: 'setDirectorySharingDevices:' is only available on macOS 12.0 or newer [-Wunguarded-availability-new]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtualMachineConfiguration.h:116:79: note: 'setDirectorySharingDevices:' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
virtualization.m:277:46: note: enclose 'setDirectorySharingDevices:' in an @available check to silence this warning
```


The code can be written this way to use an exception and to remove the warning:
```
void setDirectorySharingDevicesVZVirtualMachineConfiguration(void *config, void *directorySharingDevices)
{
    if (@available(macOS 12, *)) {
        [(VZVirtualMachineConfiguration *)config setDirectorySharingDevices:[(NSMutableArray *)directorySharingDevices copy]];
    }  else {
        [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise]; \
    }
}
```
I'd prefer to go with this code. It is very convenient and useful when the compiler can detect incompatibilities for you. It warns you when you forgot @available checks in new code.

cfergeau commented on virtualization.h:
```diff
@@ -21,7 +21,7 @@ bool shouldAcceptNewConnectionHandler(void *listener, void *connection, void *so
 /* VZVirtioSocketListener */
 @interface VZVirtioSocketListenerDelegateImpl : NSObject <VZVirtioSocketListenerDelegate>
-- (BOOL)listener:(VZVirtioSocketListener *)listener shouldAcceptNewConnection:(VZVirtioSocketConnection *)connection fromSocketDevice:(VZVirtioSocketDevice *)socketDevice;
+- (BOOL)listener:(void *)listener shouldAcceptNewConnection:(void *)connection fromSocketDevice:(void *)socketDevice;
```
This is indirectly related as this silences some warnings when building on older macOS versions. I've split the changes to a separate commit with an explanation.

Code-Hex replied:
I'm not sure what you mean. This library is supported only above macOS v11.0. So I think this is no problem.

cfergeau replied:
https://github.com/Code-Hex/vz/issues/41 asks for macOS v10 'support'. There will be no virtualization on these platforms, but errors/exceptions will be returned. Returning errors/exceptions need to be done in `Code-Hex/vz`. To add these, I use a lot compiler warnings:
```
     # github.com/Code-Hex/vz/v2
    In file included from _cgo_export.c:4:
    In file included from socket.go:6:
    ../../virtualization.h:24:19: warning: 'VZVirtioSocketListener' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
    /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtioSocketListener.h:25:12: note: 'VZVirtioSocketListener' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.13.0
    ../../virtualization.h:24:1: note: annotate 'listener:shouldAcceptNewConnection:fromSocketDevice:' with an availability attribute to silence this warning
    ../../virtualization.h:24:80: warning: 'VZVirtioSocketConnection' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
    /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtioSocketConnection.h:18:12: note: 'VZVirtioSocketConnection' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.13.0
    ../../virtualization.h:24:1: note: annotate 'listener:shouldAcceptNewConnection:fromSocketDevice:' with an availability attribute to silence this warning
    ../../virtualization.h:24:136: warning: 'VZVirtioSocketDevice' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
    /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtioSocketDevice.h:25:12: note: 'VZVirtioSocketDevice' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.13.0
    ../../virtualization.h:24:1: note: annotate 'listener:shouldAcceptNewConnection:fromSocketDevice:' with an availability attribute to silence this warning
```

If there are warnings, there is some code without `@available`. If there are pages and pages of warnings, I will not notice new code missing `@available`. This is why I prefer to remove all the warnings I can.

Code-Hex replied:
My opinion. https://github.com/Code-Hex/vz/issues/41#issuecomment-1265260233

I don't support the compilation on macOS v10. because It depends on xcode sdk.
So you don't need to consider for v10.

cfergeau replied:
These warnings happen with latest macOS/xcode when you build a binary for macOS 10 or newer:
```zsh
% sw_vers 
ProductName:	macOS
ProductVersion:	12.6
BuildVersion:	21G115

% CGO_CFLAGS=-mmacosx-version-min=10.13  make -C example/linux
go build -o virtualization .
# github.com/Code-Hex/vz/v2
In file included from virtualization.m:8:
../../virtualization_view.h:26:41: warning: 'VZVirtualMachine' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtualMachine.h:67:12: note: 'VZVirtualMachine' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.13.0
../../virtualization_view.h:26:1: note: annotate 'initWithVirtualMachine:windowWidth:windowHeight:' with an availability attribute to silence this warning
virtualization.m:1249:37: warning: 'VZVirtualMachine' is only available on macOS 11.0 or newer [-Wunguarded-availability-new]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZVirtualMachine.h:67:12: note: 'VZVirtualMachine' has been marked as being introduced in macOS 11.0 here, but the deployment target is macOS 10.13.0
virtualization.m:1249:37: note: enclose 'VZVirtualMachine' in an @available check to silence this warning
# github.com/Code-Hex/vz/v2
```

They are not related to building on macOS 10. I agree we don't need to support *building* on macOS.

Code-Hex replied:
Please revert this~

Code-Hex replied:
@cfergeau Ping

cfergeau commented on virtualization.m:
```diff
@@ -36,9 +36,13 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
 
 @implementation VZVirtioSocketListenerDelegateImpl
-- (BOOL)listener:(VZVirtioSocketListener *)listener shouldAcceptNewConnection:(VZVirtioSocketConnection *)connection fromSocketDevice:(VZVirtioSocketDevice *)socketDevice;
+- (BOOL)listener:(void *)listener shouldAcceptNewConnection:(void *)connection fromSocketDevice:(void *)socketDevice;
```
Same reasons as https://github.com/Code-Hex/vz/pull/48#discussion_r982274630

Code-Hex replied:
Please revert this~

cfergeau commented on virtualization_arm64.h:
```diff
@@ -65,7 +65,7 @@ void *newVZMacMachineIdentifierWithPath(const char *machineIdentifierPath);
 nbyteslice getVZMacMachineIdentifierDataRepresentation(void *machineIdentifierPtr);
 
-VZMacOSRestoreImageStruct convertVZMacOSRestoreImage2Struct(VZMacOSRestoreImage *restoreImage);
+VZMacOSRestoreImageStruct convertVZMacOSRestoreImage2Struct(void *restoreImagePtr);
```
This is also related to https://github.com/Code-Hex/vz/pull/48#discussion_r982274630

Code-Hex replied:
This code is used only internally. So I don't think this is no problem

Code-Hex replied:
Please revert this~

cfergeau replied:
I still think we need this ( https://github.com/Code-Hex/vz/pull/48#issuecomment-1260760961 ). It is really useful for the maintainance of macOS 11/macOS 10 compat code. If there are too many warnings, we will not notice the new ones.

It also makes the code consistent with the other `convert*` functions:
```go
virtualization.h:VZVirtioSocketConnectionFlat convertVZVirtioSocketConnection2Flat(void *connection);
virtualization_arm64.h:VZMacOSConfigurationRequirementsStruct convertVZMacOSConfigurationRequirements2Struct(void *requirementsPtr);
virtualization_arm64.h:VZMacHardwareModelStruct convertVZMacHardwareModel2Struct(void *hardwareModelPtr);
```
They all use void*

Code-Hex replied:
@cfergeau Ping
(Because I don't need to support os 10)

cfergeau replied:
This specific change is about macOS11. Commit log mentions `CGO_CFLAGS=-mmacosx-version-min=11.0` and `warning: 'VZMacOSRestoreImage' is only available on macOS 12.0 or newer`.

I've reverted the vsock one (to avoid force push). I'd like to squash all the latest changes/remove the problematic commit and then force-push though as it's getting messy ^^.

cfergeau replied:
> This function is used only internally not in the Go world.

I know it is only internal. But the objc compiler warns about it.

> (because this is not following other convert function rules).

All other convert functions are using `void *`:
- https://github.com/Code-Hex/vz/blob/37646e120f85b7fd99989e7e4b23cf068564b561/virtualization.h#L124
- https://github.com/Code-Hex/vz/blob/37646e120f85b7fd99989e7e4b23cf068564b561/virtualization_arm64.h#L72-L73

`convertVZMacOSRestoreImage2Struct` is the only one using `VZMacOSRestoreImage *` and not `void *`. This change makes all `convert` functions the same. And it removes a compilation warning. This will be useful for future maintainance.

cfergeau replied:
Warning is still present even if I remove `convertVZMacOSRestoreImage2Struct` from the header
```
# github.com/Code-Hex/vz/v2
virtualization_arm64.m:330:68: warning: 'VZMacOSRestoreImage' is only available on macOS 12.0 or newer [-Wunguarded-availability-new]
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Virtualization.framework/Headers/VZMacOSRestoreImage.h:30:12: note: 'VZMacOSRestoreImage' has been marked as being introduced in macOS 12.0 here, but the deployment target is macOS 11.0.0
virtualization_arm64.m:330:34: note: annotate 'convertVZMacOSRestoreImage2Struct' with an availability attribute to silence this warning
```

cfergeau commented:
In the latest iteration, I've added exceptions. I wrote the code as described in https://github.com/Code-Hex/vz/pull/48#discussion_r980319270
```
void setDirectorySharingDevicesVZVirtualMachineConfiguration(void *config, void *directorySharingDevices)
{
    if (@available(macOS 12, *)) {
        [(VZVirtualMachineConfiguration *)config setDirectorySharingDevices:[(NSMutableArray *)directorySharingDevices copy]];
    }  else {
        [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise]; \
    }
}
```

It is important to write it this way to avoid compiler warnings. The compiler warnings will be useful when new code is added. They will tell us about APIs which require to be wrapped in @available.

Code-Hex commented on virtualization_arm64.m:
```diff
@@ -24,16 +24,21 @@ - (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(N
-                                                                     hardwareModel:(VZMacHardwareModel *)hardwareModel
-                                                                           options:VZMacAuxiliaryStorageInitializationOptionAllowOverwrite
-                                                                             error:(NSError *_Nullable *_Nullable)error];
+    if (@available(macOS 12, *)) {
+        VZMacAuxiliaryStorage *auxiliaryStorage;
+        @autoreleasepool {
+            NSString *storagePathNSString = [NSString stringWithUTF8String:storagePath];
+            NSURL *storageURL = [NSURL fileURLWithPath:storagePathNSString];
+            auxiliaryStorage = [[VZMacAuxiliaryStorage alloc] initCreatingStorageAtURL:storageURL
+                                                                         hardwareModel:(VZMacHardwareModel *)hardwareModel
+                                                                               options:VZMacAuxiliaryStorageInitializationOptionAllowOverwrite
+                                                                                 error:(NSError *_Nullable *_Nullable)error];
+        }
+        return auxiliaryStorage;
+    } else {
+        [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise];
+        return nil;
     }
```
Could you fix like this

```objective-c
if (@available(macOS 12, *)) {
  // code
  return auxiliaryStorage; 
}
[[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise];
// should not return here because unreachable code
```

cfergeau replied:
Without an explicit `return`, there's a compiler warning:
```
warning: non-void function does not return a value in all control paths [-Wreturn-type]
```
Another option is to use `__builtin_unreachable();`:
```
if (@available(macOS 12, *)) {
  // code
  return auxiliaryStorage; 
}
[[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise];
__builtin_unreachable();
```

Code-Hex replied:
Good. It's better to make it in macro 👍

cfergeau replied:
Done.

Code-Hex commented on virtualization_arm64.m:
```diff
@@ -147,15 +187,24 @@ void setAuxiliaryStorageVZMacPlatformConfiguration(void *config, void *auxiliary
 void setDisplaysVZMacGraphicsDeviceConfiguration(void *graphicsConfiguration, void *displays)
 {
-    [(VZMacGraphicsDeviceConfiguration *)graphicsConfiguration setDisplays:[(NSMutableArray *)displays copy]];
+    if (@available(macOS 12, *)) {
+        [(VZMacGraphicsDeviceConfiguration *)graphicsConfiguration setDisplays:[(NSMutableArray *)displays copy]];
+    } else {
+        [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise];
+    }
```
better for me
```suggestion
    if (@available(macOS 12, *)) {
        [(VZMacGraphicsDeviceConfiguration *)graphicsConfiguration setDisplays:[(NSMutableArray *)displays copy]];
    }
    [[NSException exceptionWithName:@"UnhandledException" reason:@"bug" userInfo:nil] raise];
```

cfergeau replied:
I've changed this in all commits.

Code-Hex commented on virtualization.go:
```diff
@@ -7,12 +7,24 @@ package vz
 
+// ErrUnsupportedOSVersion is returned when calling a method which is only
+// available in newer macOS versions.
+// This allows for example to build an application using the file sharing API,
+// and to be able to run it both on macOS 11 and macOS 12. When running on
+// macOS 11, the file sharing API will return ErrUnsupportedOSVersion, while on
+// macOS 12, the file sharing API will work as expected.
```
```suggestion
```

cfergeau replied:
Added such an example, and shortened the doc comment.

Code-Hex commented on virtualization.go:
```diff
@@ -179,27 +198,50 @@ func (v *VirtualMachine) StateChangedNotify() <-chan VirtualMachineState {
 // CanStart returns true if the machine is in a state that can be started.
 func (v *VirtualMachine) CanStart() bool {
-	return bool(C.vmCanStart(v.Ptr(), v.dispatchQueue))
+	if macosMajorVersionLessThan(11) {
+		return false
+	} else {
+		return bool(C.vmCanStart(v.Ptr(), v.dispatchQueue))
+	}
```
Go does not recommend return within if else block
Please fix
```suggestion
	if macosMajorVersionLessThan(11) {
		return false
	}
	return bool(C.vmCanStart(v.Ptr(), v.dispatchQueue))
```

cfergeau replied:
Fixed.

Code-Hex commented:
@cfergeau If you force push, I can't check what's been fixed, so please don't do from next.

Code-Hex commented on example/linux/main.go:
```diff
@@ -56,38 +56,63 @@ func main() {
 		vz.WithInitrd(initrd),
 	)
+	if err != nil {
+		panic(err)
```
Could you replace panic with log.Fatal(err)?
I used panic because it does not complete the setup for logger in line 41.

cfergeau commented:
> @cfergeau If you force push, I can't check what's been fixed, so please don't do from next.

Ah yes, github is annoying, it does not keep old versions of PRs :(
I added `fixup` commits to this new iteration, they can be squashed before it's merged.

Code-Hex commented:
@cfergeau Thank you!
